r/R/dplyr-datetime-helpers.R

```diff
@@ -201,19 +213,130 @@ build_formats <- function(orders) {
 }
 
 build_format_from_order <- function(order) {
-  year_chars <- c("%y", "%Y")
-  month_chars <- c("%m", "%B", "%b")
-  day_chars <- "%d"
-
-  outcome <- switch(
-    order,
-    "ymd" = expand.grid(year_chars, month_chars, day_chars),
-    "ydm" = expand.grid(year_chars, day_chars, month_chars),
-    "mdy" = expand.grid(month_chars, day_chars, year_chars),
-    "myd" = expand.grid(month_chars, year_chars, day_chars),
-    "dmy" = expand.grid(day_chars, month_chars, year_chars),
-    "dym" = expand.grid(day_chars, year_chars, month_chars)
+  char_list <- list(
+    "y" = c("%y", "%Y"),
+    "m" = c("%m", "%B", "%b"),
+    "d" = "%d",
+    "H" = "%H",
+    "M" = "%M",
+    "S" = "%S"
+  )
+
+  split_order <- strsplit(order, split = "")[[1]]
+
+  outcome <- expand.grid(char_list[split_order])
+  formats_with_sep <- do.call(paste, c(outcome, sep = "-"))
+  formats_without_sep <- do.call(paste, c(outcome, sep = ""))
+  c(formats_with_sep, formats_without_sep)
```

Comment:
This would be something we would want to measure with a benchmark on a bunch of data, but something to consider is instead of doing both with and without separators on all, you could do something like `ifelse(grepl("-", col), formats_with_sep, formats_without_sep)` — though not quite right here, since you would want that to be passed as an expression

Reply:
What is the data has a mixture of both? I imagine it could happen in very messy datasets. So probably more like:
```r
ifelse(grepl("-", col), try_both_formats, try_only_without_sep) 
```

Reply:
`ifelse(grepl(...), ...)` should be operating at the element level for the column (e.g. it should be vectorized) so that there might be a mixture shouldn't be a problem.

Reply:
Using `ifelse` conditional on `x` containing a `"-"` returns an Expression. Currently I am using `purrr::map()` to build a list of Expressions, which would not work in this instance.

This is a sketch of the logic showing where the `ifelse` approach would trip up.
```r
formats <- call_binding(
  "if_else",
  call_binding("grepl", "-", x),
  formats_with_sep,
  formats_without_sep
  )

# now formats is an Expression and we cannot map over it; this would fail
  map(
    formats,
    ~ build_expr(
      "strptime",
      x,
      options = list(format = .x, unit = 0L, error_is_null = TRUE)
    )
  )
```

Reply:
In conclusion, I propose `formats` stays as-is until we have a way to iterate over Expressions.

Reply:
Right, you would need to add that `if_else()` to the expressions. This will require this code be in a slightly different location to take be added in the right spot (though you should be able to construct a test to see what kinds of performance impacts there are for running 1x orders versus 2x in this case).

Reply:
I managed to apply the essence of this, but in a slightly different way: 
* return a single version of `formats` - with separators
* map over them, but remote the separator if `x` doesn't have one

Reply:
@jonkeane I'm going to mark this as **done**. I haven't run any tests, but the approach makes a lot of sense. Happy to have a follow-up ticket and see if this move actually had a positive impact on performance. I don't expect it to have a negative impact though.